src/main/java/io/appform/dropwizard/sharding/BundleCommonBase.java

```diff
@@ -15,66 +14,96 @@
-    validateAndBuildEntitiesMeta(initialisedEntities);
+  protected BundleCommonBase(final Class<?> entity, final Class<?>... entities) {
+    initialiseEntitiesAndMeta(ImmutableList.<Class<?>>builder()
+            .add(entity).add(entities).build());
   }
 
-  protected BundleCommonBase(List<String> classPathPrefixList) {
-    Set<Class<?>> entities = new Reflections(classPathPrefixList).getTypesAnnotatedWith(
-        Entity.class);
+  protected BundleCommonBase(final List<String> classPathPrefixList) {
+    final var entities = new Reflections(classPathPrefixList).getTypesAnnotatedWith(Entity.class);
     Preconditions.checkArgument(!entities.isEmpty(),
-        String.format("No entity class found at %s",
-            String.join(",", classPathPrefixList)));
-    this.initialisedEntities = ImmutableList.<Class<?>>builder().addAll(entities).build();
-    validateAndBuildEntitiesMeta(initialisedEntities);
+            String.format("No entity class found at %s", String.join(",", classPathPrefixList)));
+    initialiseEntitiesAndMeta(entities);
+  }
+
+  public final void registerEntities(@NonNull final Class<?>... entities) {
+      initialisePendingInitialisationEntitiesAndMeta(Arrays.asList(entities));
+  }
+
+  public final void registerEntities(@NonNull final List<String> classPathPrefixList) {
+      final var entities = new Reflections(classPathPrefixList).getTypesAnnotatedWith(Entity.class);
+      Preconditions.checkArgument(!entities.isEmpty(),
+              String.format("No entity class found at %s", String.join(",", classPathPrefixList)));
+      initialisePendingInitialisationEntitiesAndMeta(entities);
+  }
+
+  private synchronized void initialisePendingInitialisationEntitiesAndMeta(final Collection<Class<?>> entities) {
+    if (entityInitialisationBlocked) {
+      throw new UnsupportedOperationException("Entity registration is not supported after run method execution.");
+    }
+    final var pendingInitialisationEntities = entities.stream()
+            .filter(entity -> !this.initialisedEntities.contains(entity))
+            .collect(Collectors.toList());
+    initialiseEntitiesAndMeta(pendingInitialisationEntities);
+  }
+
+  private void initialiseEntitiesAndMeta(final Collection<Class<?>> entities) {
+    this.initialisedEntities.addAll(entities);
+    validateAndBuildEntitiesMeta(entities);
+  }
+
+  protected final synchronized void blockEntityInitialisation() {
+    this.entityInitialisationBlocked = true;
   }
 
   protected ShardBlacklistingStore getBlacklistingStore() {
     return new NoopShardBlacklistingStore();
   }
 
   public List<Class<?>> getInitialisedEntities() {
-    if (this.initialisedEntities == null) {
+    if (this.initialisedEntities.isEmpty()) {
       throw new IllegalStateException("DB sharding bundle is not initialised !");
     }
-    return this.initialisedEntities;
+    return List.copyOf(this.initialisedEntities);
```

Comment:
What's the intent of this change - Do we want a read only copy of existing list or a snapshot of existing entities list. Current implementation (List.copyOf()) will create a new list every single time which could hamper performance (if existing clients are using this method frequently)